File: algorithms/04_java/graphs/MaxFlowDinic.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Deque;
import java.util.List;

public class MaxFlowDinic {
    static class Edge {
        int to, rev;
        long cap;
        Edge(int to, long cap, int rev) { this.to = to; this.cap = cap; this.rev = rev; }
    }

    private List<Edge>[] g;
    private int[] level, it;

    @SuppressWarnings("unchecked")
    public MaxFlowDinic(int n) {
        g = new List[n];
        for (int i = 0; i < n; i++) g[i] = new ArrayList<>();
        level = new int[n];
        it = new int[n];
    }

    public void addEdge(int u, int v, long cap) {
        g[u].add(new Edge(v, cap, g[v].size()));
        g[v].add(new Edge(u, 0, g[u].size() - 1));
    }

    private boolean bfs(int s, int t) {
        Arrays.fill(level, -1);
        Deque<Integer> q = new ArrayDeque<>();
        level[s] = 0;
        q.add(s);
        while (!q.isEmpty()) {
            int u = q.poll();
            for (Edge e : g[u]) {
                if (e.cap > 0 && level[e.to] == -1) {
                    level[e.to] = level[u] + 1;
                    q.add(e.to);
                }
            }
        }
        return level[t] != -1;
    }

    private long dfs(int u, int t, long f) {
        if (u == t) return f;
        for (; it[u] < g[u].size(); it[u]++) {
            Edge e = g[u].get(it[u]);
            if (e.cap > 0 && level[e.to] == level[u] + 1) {
                long d = dfs(e.to, t, Math.min(f, e.cap));
                if (d > 0) {
                    e.cap -= d;
                    g[e.to].get(e.rev).cap += d;
                    return d;
                }
            }
        }
        return 0;
    }

    public long maxFlow(int s, int t) {
        long flow = 0;
        while (bfs(s, t)) {
            Arrays.fill(it, 0);
            long f;
            while ((f = dfs(s, t, Long.MAX_VALUE)) > 0) flow += f;
        }
        return flow;
    }

    public static void main(String[] args) {
        MaxFlowDinic dinic = new MaxFlowDinic(6);
        dinic.addEdge(0, 1, 16);
        dinic.addEdge(0, 2, 13);
        dinic.addEdge(1, 2, 10);
        dinic.addEdge(1, 3, 12);
        dinic.addEdge(2, 1, 4);
        dinic.addEdge(2, 4, 14);
        dinic.addEdge(3, 2, 9);
        dinic.addEdge(3, 5, 20);
        dinic.addEdge(4, 3, 7);
        dinic.addEdge(4, 5, 4);
        long flow = dinic.maxFlow(0, 5);
        if (flow != 23) throw new AssertionError("max flow should be 23, got " + flow);
        System.out.println("[Java MaxFlowDinic] Dinic's max flow verified: " + flow);
    }
}

File: algorithms/06_javascript/graphs/max_flow_dinic.js
class Dinic {
  constructor(n) {
    this.n = n;
    this.graph = Array.from({ length: n }, () => []);
  }
  addEdge(u, v, cap) {
    this.graph[u].push({ to: v, cap, rev: this.graph[v].length });
    this.graph[v].push({ to: u, cap: 0, rev: this.graph[u].length - 1 });
  }
  bfs(s, t) {
    this.level = new Array(this.n).fill(-1);
    this.level[s] = 0;
    const q = [s];
    while (q.length) {
      const u = q.shift();
      for (const e of this.graph[u]) {
        if (e.cap > 0 && this.level[e.to] < 0) {
          this.level[e.to] = this.level[u] + 1;
          q.push(e.to);
        }
      }
    }
    return this.level[t] >= 0;
  }
  dfs(u, t, f) {
    if (u === t) return f;
    for (; this.it[u] < this.graph[u].length; this.it[u]++) {
      const e = this.graph[u][this.it[u]];
      if (e.cap > 0 && this.level[e.to] === this.level[u] + 1) {
        const d = this.dfs(e.to, t, Math.min(f, e.cap));
        if (d > 0) {
          e.cap -= d;
          this.graph[e.to][e.rev].cap += d;
          return d;
        }
      }
    }
    return 0;
  }
  maxFlow(s, t) {
    let flow = 0;
    const INF = Infinity;
    while (this.bfs(s, t)) {
      this.it = new Array(this.n).fill(0);
      let f;
      while ((f = this.dfs(s, t, INF)) > 0) flow += f;
    }
    return flow;
  }
}

module.exports = { Dinic };

if (require.main === module) {
  const dinic = new Dinic(4);
  dinic.addEdge(0, 1, 3);
  dinic.addEdge(0, 2, 2);
  dinic.addEdge(1, 2, 1);
  dinic.addEdge(1, 3, 2);
  dinic.addEdge(2, 3, 2);
  if (dinic.maxFlow(0, 3) !== 4) throw new Error("dinic max flow failed");
  console.log("[JavaScript Dinic Max Flow] Dinic's algorithm verified");
}
